roles/jd-client/src/lib/channel_manager/jd_message_handler.rs

```
@@ -0,0 +1,218 @@
+                .send(frame.into())
+                .await
+                .map_err(|_e| Error::External(Box::new(JDCError::ChannelErrorSender)))?;
+
+            info!("Sent updated CoinbaseOutputConstraints to TP channel");
+        } else {
+            debug!("Coinbase outputs unchanged, skipping constraints update");
+        }
+
+        Ok(())
+    }
+
+    #[instrument(skip_all, fields(request_id = msg.request_id), level="trace")]
+    async fn handle_declare_mining_job_error(
+        &mut self,
+        msg: DeclareMiningJobError<'_>,
+    ) -> Result<(), Error> {
+        warn!(
+            error_code = ?msg.error_code,
+            "Received DeclareMiningJobError from JDS"
+        );
+        debug!("AllocateMiningJobTokenSuccess: {msg:?}");
+
+        debug!("Fallback path triggered for request_id={}", msg.request_id);
+        _ = self
+            .channel_manager_channel
+            .status_sender
+            .send(Status {
+                state: State::JobDeclaratorShutdownFallback(JDCError::Shutdown),
+            })
+            .await;
+
+        Ok(())
+    }
+
+    #[instrument(skip_all, fields(request_id = msg.request_id), level="trace")]
+    async fn handle_declare_mining_job_success(
+        &mut self,
+        msg: DeclareMiningJobSuccess<'_>,
+    ) -> Result<(), Error> {
+        info!("Received DeclareMiningJobSuccess from JDS");
+        debug!("DeclareMiningJobSuccess: {msg:?}");
+
+        let Some(last_declare_job) = self
+            .channel_manager_data
+            .super_safe_lock(|data| data.last_declare_job_store.get(&msg.request_id).cloned())
+        else {
+            warn!(
+                "No last_declare_job found for request_id={}",
+                msg.request_id
+            );
+            return Ok(());
+        };
+
+        let Some(custom_job) = last_declare_job.custom_job else {
+            debug!("No custom_job present for request_id={}", msg.request_id);
+            return Ok(());
```

Shouldn't we return a specific error for this?

---

Also, who is in charge of setting the custom_job field?

And when is it set?